### ISSUE TITLE
CustomIntegrator bug on OpenCL and CUDA

### ISSUE THREAD

**Issue by jchodera** [pandegroup/openmm#1380]
There is a bug in `CustomIntegrator` that does not transfer the value of accumulated sums to their variables appearing in subsequent global variable expressions on `OpenCL` and `CUDA` (but not `Reference`).

Minimum test example:
```python
from simtk import openmm, unit

class TestIntegrator(openmm.CustomIntegrator):
    def __init__(self, timestep):
        super(TestIntegrator, self).__init__(timestep)

        self.addGlobalVariable('kinetic', 0.0) # kinetic energy
        self.addGlobalVariable('initial_total_energy', 0.0) # initial total energy (kinetic + potential)
        self.addGlobalVariable('final_total_energy', 0.0) # final total energy (kinetic + potential)
        self.addGlobalVariable('log_ncmc_acceptance_probability', 0.0) # log of NCMC acceptance probability
        self.addGlobalVariable('dti', timestep.value_in_unit_system(unit.md_unit_system)) # inner timestep
        self.addGlobalVariable('lambda', 0.0) # parameter switched from 0 <--> 1 during course of integrating internal 'nsteps' of dynamics
        self.addPerDofVariable("x1", 0) # for velocity Verlet with constraints

        # Constrain initial positions and velocities.
        self.addConstrainPositions()
        self.addConstrainVelocities()
        self.addUpdateContextState()

        # Store initial total energy.
        self.addComputeSum("kinetic", "0.5*m*v*v")

        # Retrieve kinetic energy.
        self.addComputeGlobal('log_ncmc_acceptance_probability', 'kinetic') # DEBUG


system = openmm.System()
system.addParticle(12.0)

platform = openmm.Platform.getPlatformByName("OpenCL")
integrator = TestIntegrator(1.0*unit.femtoseconds)
context = openmm.Context(system, integrator, platform)
context.setPositions([[0, 0, 0]])
context.setVelocities([[1, 1, 1]])

integrator.step(1)

kinetic = integrator.getGlobalVariableByName('log_ncmc_acceptance_probability')
print kinetic
```
On `OpenCL` and `CUDA`, the result is `0.0`, but on `Reference`, the correct result of `18.0` is returned.

cc #1379 and https://github.com/choderalab/perses/pull/85 where I said my integrator gave really weird statistics.  This is why.

**Comment by jchodera** [comment]
Here's a trimmed-down example:
```python
from simtk import openmm, unit

class TestIntegrator(openmm.CustomIntegrator):
    def __init__(self, timestep):
        super(TestIntegrator, self).__init__(timestep)

        self.addGlobalVariable('kinetic', 0.0)
        self.addGlobalVariable('global', 0.0) 

        self.addComputeSum("kinetic", "0.5*m*v*v")
        self.addComputeGlobal('global', 'kinetic')


system = openmm.System()
system.addParticle(12.0)

for platform_index in range(openmm.Platform.getNumPlatforms()):
    platform = openmm.Platform.getPlatform(platform_index)
    platform_name = platform.getName()
    integrator = TestIntegrator(1.0*unit.femtoseconds)
    context = openmm.Context(system, integrator, platform)
    context.setPositions([[0, 0, 0]])
    context.setVelocities([[1, 1, 1]])
    integrator.step(1)
    print "%12s : %12.6f" % (platform_name, integrator.getGlobalVariableByName('global'))
    del context, integrator
```
Output:
```
mski1776:perses.choderalab choderaj$ python customintegrator_bug.py
   Reference :    18.000000
         CPU :    18.000000
      OpenCL :     0.000000
```

**Comment by rmcgibbo** [comment]
Looks broken to me:
```
$ python test.py
  File "test.py", line 25
    print "%12s : %12.6f" % (platform_name, integrator.getGlobalVariableByName('global'))
                        ^
SyntaxError: invalid syntax
```
:-)

**Comment by jchodera** [comment]
Fixed, just for you young python 3.x whippersnappers.

**Comment by rmcgibbo** [comment]
Okay, on my MBP laptop (Iris embedded GPU), I did

```
$ conda install -y openmm
$ python test.py
   Reference :    18.000000
         CPU :    18.000000
      OpenCL :    17.999998

$ conda install -y openmm-dev
$ python test.py
   Reference :    18.000000
         CPU :    18.000000
      OpenCL :     0.000000
```

**Comment by jchodera** [comment]
Someone broked it?

**Comment by jchodera** [comment]
Looking for potential changes to `OpenCLKernels.cpp` that may have caused this (assuming it is a code error, not a compilation environment issue).

[This one](https://github.com/pandegroup/openmm/commit/c744444af536c2ea6cf00540d88cbf3ef9d1302f) looks suspicious:
```python
         // Identify steps that can be merged into a single kernel.
          
          for (int step = 1; step < numSteps; step++) {
 -            if (invalidatesForces[step] || ((needsForces[step] || needsEnergy[step]) && forceGroupFlags[step] != forceGroupFlags[step-1]))
 +            if ((needsForces[step] || needsEnergy[step]) && (invalidatesForces[step-1] || forceGroupFlags[step] != forceGroupFlags[step-1]))
                  continue;
              if (stepType[step-1] == CustomIntegrator::ComputePerDof && stepType[step] == CustomIntegrator::ComputePerDof)
                  merged[step] = true;
```

**Comment by rmcgibbo** [comment]
As an aside -- really this is ancillary to the main point of this thread:
 - The natural thing to do here, for me, would be to use git-bisect.
 - But one of the problems with git-bisect for OpenMM is that it takes a long time to compile.
 - conda `openmm-dev` isn't particularly useful here because it only has the latest build, but not the intermediate ones
 - I don't know what the costs are, but it might not be a bad idea to store all of the builds of openmm-dev, instead of just the latest. I know this is what psi4 does (although I don't know how often they rebuild, it might not be every night). https://github.com/psi4/psi4public/blob/master/conda-recipe/psi4/meta.yaml, http://anaconda.org/psi4/psi4/files

**Comment by rmcgibbo** [comment]
lol the disincentive against adding anything fancy with the caching of pre-compiled binaries is that, as always, @peastman can fix this stuff in like 10 seconds, long before I could even look up the documentation on git bisect :-)

**Comment by peastman** [comment]
It's very rare that I have to resort to bisecting to find a bug.  It happens, but only once every several months, and when I do I usually just do the bisecting by hand, specifically looking for revisions that seem likely to have affected the behavior in question.  In this case, with such a good test case, it only took me a few minutes to track down the problem.

**Comment by jchodera** [comment]
> I don't know what the costs are, but it might not be a bad idea to store all of the builds of openmm-dev, instead of just the latest. I know this is what psi4 does (although I don't know how often they rebuild, it might not be every night).

+100!

I also suggested this [here](https://github.com/pandegroup/openmm/issues/526) because there are other reasons this is useful.
